refactor: Forbid calling unsafe fs::path(std::string) constructor and fs::path::string() method

ryanofsky opened the pull request:
The `fs::path` class has a `std::string` constructor which will implicitly convert from strings. Implicit conversions like this are not great in general because they can hide complexity and inefficiencies in the code, but this case is especially bad, because after the transition from `boost::filesystem` to `std::filesystem` in #20744 the behavior of this constructor on windows will be more complicated and can mangle path strings. The `fs::path` class also has a `.string()` method which is inverse of the constructor and has the same problems.

Fix this by replacing the unsafe method calls with `PathToString` and `PathFromString` function calls, and by forbidding unsafe method calls in the future.

ryanofsky committed:
refactor: Block unsafe fs::path std::string conversion calls

There is no change in behavior. This just helps prepare for the
transition from boost::filesystem to std::filesystem by avoiding calls
to methods which will be unsafe after the transaction to std::filesystem
to due lack of a boost::filesystem::path::imbue equivalent and inability
to set a predictable locale.

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>
Co-authored-by: Kiminuo <kiminuo@protonmail.com>
Co-authored-by: MarcoFalke <falke.marco@gmail.com>

ryanofsky committed:
refactor: Add fs::PathToString, fs::PathFromString, u8string, u8path functions

There is no change in behavior. This just helps prepare for the
transition from the boost::filesystem to the std::filesystem path
implementation.

Co-authored-by: Kiminuo <kiminuo@protonmail.com>

MarcoFalke commented on src/addrdb.cpp:
```diff
@@ -114,16 +114,16 @@ bool DeserializeFileDB(const fs::path& path, Data& data, int version)
     CAutoFile filein(file, SER_DISK, version);
     if (filein.IsNull()) {
-        LogPrintf("Missing or invalid file %s\n", path.string());
+        LogPrintf("Missing or invalid file %s\n", fs::PathToString(path));
```
Wasn't `path.string()` already explicit enough? Also, when should `u8string` be used instead? Finally, `strprintf("%s", path)` without any helpers still compiles? When should/shouldn't that be used?

ryanofsky replied:
> Wasn't `path.string()` already explicit enough?

The problem with `path.string()` isn't the explicitness, it's the unsafe behavior with the new path class on windows, where it returns a string in an arbitrary encoding based on the local code page. In preparation for this, this PR is disabling the `path.string()` now so it isn't misused later, and adding `PathToString` as a replacement which preserves the behavior we need and is complementary to `PathFromStrong`.

> Also, when should `u8string` be used instead?

`u8string/u8path` functions should be used for converting strings/paths where the strings must be valid UTF-8 strings. `PathToString/PathFromString` functions should be used otherwise.

U8 functions are the right functions to call in JSON-RPC code and Qt string code because Qt strings and JSON requests/responses require valid UTF-8. But otherwise `PathToString/PathFromString` functions which pass along strings verbatim and avoid making encoding assumptions are more appropriate. `PathToString/PathFromString` functions are the right functions to call for parsing argv paths, parsing environment variable paths, parsing config paths, writing to log files, and calling POSIX APIs that expect and return native 8-bit path strings. In all these places UTF-8 encoding is typical and nice to have, but not something that should be enforced or required. If someone has their home directory set to a non-utf8 path, `PathToString/PathFromString` won't mangle the string or care about the encoding.

Everything above is actually documented in the followup commit 3e88c9f6b4e84ab5da78f4862266b5ff13bb8fc0 ([tag](https://github.com/ryanofsky/bitcoin/commits/review.20744.8-edit.1)) posted https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-916627496, but I threw together the documentation really quickly while trying to converge on a solution, so it could use some cleanup and reshuffling, either in this PR or the std::filesystem PR later when these functions stop being identical stubs.

> Finally, `strprintf("%s", path)` without any helpers still compiles? When should/shouldn't that be used?

I can look into that because I'm not too familiar with tinyformat, but I would suspect that this always compiled and never worked at runtime maybe? I'm not sure how it could have worked unless tinyformat knew about boost types.

MarcoFalke replied:
I think tinyformat just "emulates" our `::ToString(path)`, which should also compile (didn't try)

ryanofsky replied:
> I think tinyformat just "emulates" our `::ToString(path)`, which should also compile (didn't try)

Apparently tinyformat is using operator<<, which boost path and std path both provide. I found I could disable it with

```c++
template<class CharT, class Traits>
std::basic_ostream<CharT,Traits>& operator<<(std::basic_ostream<CharT,Traits>& os, const path& p)
{
    static_assert(sizeof(CharT) == -1, "Implicit path to string conversion is error-prone and not supported, please use explicit PathToString call");
    return os;
}
```

But I found that this also broke some boost tests, and also I'm a little reluctant to make this change anyway since formatting support seems convenient for debugging. I'm leaving this alone for now but happy to do whatever people prefer here.

MarcoFalke replied:
> PathToString/PathFromString functions should be used otherwise
> U8 functions are the right functions to call in JSON-RPC code and Qt string code ...


I am not sure how this can be achieved, without something like bilingual_str.

For example, `InitError`s are passed to Qt, as well as stderr, as well as the debug log. If Qt requires `u8string` and the others require `PathToString`, then with the current approach here this is impossible to achieve, no?

ryanofsky replied:
Nothing we want to achieve should be impossible to achieve or require a different string type. If your home directory is `"/home/\xf0"` (invalid utf-8), and Qt needs to display an error message about peers.dat, it should just display it with a [replacement character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character) like "Error reading /home/�/.bitcoin/peers.dat". If we want to return the same file path as JSON-RPC, we might choose a different escaping mechanism like [python's lone surrogates](https://www.python.org/dev/peps/pep-0383/) `"/home/\udcf0/.bitcoin/peers.dat"`.

At the boundaries between our application and external programs and libraries like JSON-RPC and Qt, different ways of representing paths as strings will be necessary. That is why standard encoding functions like `u8path` and `u8string` are useful, and why we will choose different encoding functions in different places we are interacting with external code. But internally, any path can always be represented as a `path` object, and any path can always be represented as a string of bytes, as long as we are consistent about how we interpret those bytes. The `PathToString` and `PathFromString` function provide a consistent way to represent path objects as strings of bytes, now that we will no longer be able to rely on the `boost::filesystem::path` class's string methods. (The new `std::filesystem::path` class has similar methods, but unlike boost's methods their behavior cannot be controlled with an imbue function, so they are unsafe and expose complicated operating system specific behavior and will mangle strings, which is why this PR avoids and prevents calling them.)

hebasto commented:
mingw complains:
```
wallet/rpcdump.cpp: In lambda function:
wallet/rpcdump.cpp:553:76: error: use of deleted function ‘fs::path::path(const string&)’
  553 |         file.open(request.params[0].get_str(), std::ios::in | std::ios::ate);
      |                                                                            ^
In file included from ./flatfile.h:11,
                 from ./chain.h:11,
                 from wallet/rpcdump.cpp:5:
./fs.h:40:5: note: declared here
   40 |     path(const std::string& string) = delete;
      |     ^~~~
make[2]: *** [Makefile:10361: wallet/libbitcoin_wallet_a-rpcdump.o] Error 1
```

ryanofsky reviewed (commented):
Rebased 1e071392799def86e1cb6d93e83b985df025f0f8 -> ddbfc51002d3ef3238bc796b1aec159f361828b9 ([`pr/u8path.1`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.1) -> [`pr/u8path.2`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.1-rebase..pr/u8path.2)) due to conflict with #21222, also fixing win64 build error https://cirrus-ci.com/task/5157617598201856

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-916814941

> mingw complains:

Thanks for headsup, and this should be addressed in the last push

ryanofsky commented:
This PR is basically ready but I still am looking for feedback about about whether best solution to `path.string()` method and `path(string)` constructor brokenness on windows is to make these methods uncallable and replace with them with `PathToString` `PathFromString` functions like this PR is doing. Or to just override these methods to keep current boost path behavior, and actually do what we want, but not follow the standard path class behavior. Or to take an alternate approach like using the new (since mid-2019 windows build option) to force a UTF-8 code page, and not worry about older windows versions.

---

Other remaining todos here: improve fs.h documentation and split commit into two parts so fs.h functions are added in the first commit, and code is switched over to call the new functions in a second commit.

MarcoFalke commented:
> Or to take an alternate approach like using the new (since mid-2019) windows build option to force a UTF-8 code page

If this fixes all our issues and avoids developer headaches when writing code (which conversion function to use), that seems most preferable.

ryanofsky commented:
> > Or to take an alternate approach like using the new (since mid-2019) windows build option to force a UTF-8 code page
> 
> If this fixes all our issues and avoids developer headaches when writing code (which conversion function to use), that seems most preferable.

Sorry, I thought I had posted more details/drawbacks about this alternate approach previously, but I don't think I ever did. Details can be found at https://docs.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page#set-a-process-code-page-to-utf-8

The only impact I believe this alternate approach would have for future developers (as opposed to current reviewers) compared to the current approach is that future developers will be able to write `p.string()` instead of `PathToString(p)` and write `fs::path(s)` instead of `fs::PathFromString(s)`. And I think this difference is actually a drawback, not a benefit of that approach, because the `fs::path` constructor is not explicit, and the implicit conversions can lead to unintended bugs that are less likely to be cause by reviewer. Also I think it is better to have our own `PathToString` and `PathFromString` functions because these give a way to document more obviously which functions should be used for our project, instead of leaving developers looking at abstract cppreference pages that do not give concrete information about what the functions actually do on different platforms, or give any guidance on whether it is appropriate to chose `.string()` or `.u8string()` or `.native()` conversion methods in the context of writing bitcoin application code.

Another drawback of the UTF-8 windows code page build option is that is silently ignored by older versions of windows.

Another drawback of the UTF-8 windows code page build option is that it is not yet unimplemented, and will require someone with more knowledge than me of both the MSVC and Mingw builds to implement, while this PR is a straightforward code change that is basically ready to go.

hebasto commented:
Out of curiosity, what is the leveldb approach to handle paths on Windows?

hebasto commented:
friendly ping @sipsorcery

ryanofsky commented:
> friendly ping @sipsorcery

:facepalm: To be clear I think that the build hack would be inferior to this PR in every way! I think think the build hack is not worth implementing or even discussing seriously. The only reason I brought it up was for completeness and to illustrate how fragile and unpredictable the [`std::filesystem::path(std::string)` implicit constructor](https://en.cppreference.com/w/cpp/filesystem/path/path) and [`std::filesystem::path::string()` method](https://en.cppreference.com/w/cpp/filesystem/path/string) are, and why I think we should forbid them in favor of explicit `PathFromString/PathToString` functions that we can call straightforwardly and explain straightforwardly, to preserve correct behavior of `boost::filesystem::path`, and avoid a small pitfall in `std::filesystem::path`!

> Out of curiosity, what is the leveldb approach to handle paths on Windows?

leveldb just uses string paths, not `boost::filessystem::path` or `std::filesystem::path`. Handling windows paths is really not that complicated. All you have to do is avoid older code page APIs and use newer APIs that ignore the windows code page. In the new `std::filesystem` there are two methods that are using the windows code page, instead of ignoring it like we want, and like the `boost::filesystem` equivalents did. So this PR is just switching code to avoid these two particular methods, and prevent them from being called in the future.

hebasto commented:
Concept ACK on adding own explicit path <-> string conversion functions.

> To be clear I think that the build hack would be inferior to this PR in every way! I think think the build hack is not worth implementing or even discussing seriously.

Sorry Russ, but I don't follow what "the build hack" do you mean.

> > Out of curiosity, what is the leveldb approach to handle paths on Windows?
> 
> leveldb just uses string paths, not `boost::filessystem::path` or `std::filesystem::path`. Handling windows paths is really not that complicated. All you have to do is avoid older code page APIs and use newer APIs that ignore the windows code page. In the new `std::filesystem` there are two methods that are using the windows code page, instead of ignoring it like we want, and like the `boost::filesystem` equivalents did. So this PR is just switching code to avoid these two particular methods, and prevent them from being called in the future.

Thanks for explanation!

hebasto commented:
Do I understand correctly that all changes (except for `fs.h` and `fs.cpp`) are "forced" by a compiler due to https://github.com/bitcoin/bitcoin/blob/d312e529ce904c0dc8244f29424b8cba9c7da9cd/src/fs.h#L40-L42?

ryanofsky commented:
> Do I understand correctly that _all_ changes (except for `fs.h` and `fs.cpp`) are "forced" by a compiler

Yes that is exactly right. I'm working on splitting up the PR into commits and rewriting the description and documentation now that various test bugs have been fixed, but this PR is just forbidding calls to `fs::path` methods that work ok on unix, but are confusing, badly documented, and can mangle strings on windows. It is replacing those method calls with `PathToString/PathFromString` calls and `u8string/u8path` calls which all do the same thing in the `boost::filesystem` implemenation, but will avoid complexity and bugs after the`std::filesystem` switch.

sipsorcery commented:
@hebasto this stuff is a bit above my pay grade I'm afraid. I'm ok with the Win32 API's, such as `CreateFileA` for ANSI, `CreateFileW` for Unicode (UTF-16) etc. but when it comes to the file paths in the standard library `std::filesystem` and `boost` I'm out of my depth.

It's a pity @ken2812221 isn't still around, he seemed to have a good grip on this stuff. I've always wondered where he got to?

https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+author%3Aken2812221+unicode

ryanofsky commented:
I'm working on improving documentation so this will be easier to understand, but the idea is just that we need to call functions that convert strings predictably and ignore the current windows "code page". We need to avoid calling functions that convert strings unpredictably and use the the windows code page.

The new `fs::path` class happens to have two methods which we are calling which ignore the windows code page currently in the `boost::filesystem` implementation but unfortunately will start using the windows code page after #20744 which switches to the `std::filesystem` implementation.

So this PR is just forbidding calls to those two methods and replacing them with other calls.

hebasto commented on src/fs.h:
```diff
@@ -15,7 +15,55 @@
+    using boost::filesystem::path::path;
+    path(boost::filesystem::path path) : boost::filesystem::path::path(std::move(path)) {}
+    path(const std::string& string) = delete;
+    path& operator=(std::string&) = delete;
```
I think we should also add
```suggestion
    path& operator=(std::string&) = delete;
    path& operator/=(const std::string&) = delete;
    path& append(const std::string&) = delete;
```

MarcoFalke commented:
@ryanofsky Have you seen my previous comment https://github.com/bitcoin/bitcoin/pull/22937#discussion_r706099389 ?

MarcoFalke commented on src/fs.cpp:
```diff
@@ -249,9 +249,9 @@ void ofstream::close()
+static_assert(sizeof(*BOOST_FILESYSTEM_C_STR(boost::filesystem::path())) == sizeof(wchar_t),
 #else
-static_assert(sizeof(*fs::path().BOOST_FILESYSTEM_C_STR) == sizeof(wchar_t),
+static_assert(sizeof(*boost::filesystem::path().BOOST_FILESYSTEM_C_STR) == sizeof(wchar_t),
```
Can probably remove all of this code, because `BOOST_FILESYSTEM_C_STR` is no longer used? If yes, maybe even split up this change

fanquake replied:
I'm happy to base #20744 on this change, if this is the direction we are taking, and save splitting the Boost removal over several PRs.

MarcoFalke replied:
If we want to move forward with removing `boost::fs`, it could even make sense to split up the build system changes, so that other pull requests can start using the other C++17 features.

ryanofsky replied:
> I'm happy to base #20744 on this change, if this is the direction we are taking, and save splitting the Boost removal over several PRs.

Yeah, sorry, I've been a slow iterating over the documentation and variations on the idea, but I'll post an updated #20744 and rebased version of this PR today

MarcoFalke commented on src/fs.h:
```diff
@@ -15,7 +15,55 @@
+    path(const std::string& string) = delete;
+    path& operator=(std::string&) = delete;
+    std::string string() const = delete;
+    std::string u8string() const { return boost::filesystem::path::string(); }
```
Would it make sense to make `operator<<(a)` call `a<<PathToString(*this)`?

ryanofsky reviewed (commented):
Updated d312e529ce904c0dc8244f29424b8cba9c7da9cd -> f735d6031d827e8a1bea5dcb00a11262be9c8a20 ([`pr/u8path.3`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.3) -> [`pr/u8path.4`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.3..pr/u8path.4)) with suggested changes, also splitting commits and improving documentation

hebasto commented on src/fs.h:
```diff
@@ -15,7 +15,109 @@
+public:
+    using boost::filesystem::path::path;
+
+    // Allow path objects arguments for comptability.
```
typo: 
```suggestion
    // Allow path objects arguments for compatibility,.
```

hebasto reviewed (commented):
Looks great!

Functional tests still [fail](https://api.cirrus-ci.com/v1/task/6544866634104832/logs/functional_tests.log) on Windows for the [combined branch](https://github.com/ryanofsky/bitcoin/commits/pr/stdfs.4).

ryanofsky commented:
Updated f735d6031d827e8a1bea5dcb00a11262be9c8a20 -> 650470227f9312b627b4bee3f85540f2e00f8bd7 ([`pr/u8path.4`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.4) -> [`pr/u8path.5`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.4..pr/u8path.5)) fixing review comment about bad spelling, and fixing error in multiprocess build https://cirrus-ci.com/task/4610801555210240?logs=ci#L3269

hebasto reviewed (approved):
ACK 650470227f9312b627b4bee3f85540f2e00f8bd7

I tested this PR by building #20744 on top of this one. In the combined branch functional test on Windows (MSVC build) fail due to the wrong encoding. They could be fixed by adding `std::setlocale` into the `SetupEnvironment` function (just a demo, not a suggestion for the real code). ~But I'm sure these encoding issues are out of this PR scope, and they belong to the `fsbridge` namespace (see #13866).~ See https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-919610206

UPDATE
I continued testing of the combined branch (#20744 on top of this PR), and found another function with unwanted implicit conversion which must be handled:

```diff
--- a/src/fs.h
+++ b/src/fs.h
@@ -59,6 +59,11 @@ static inline path operator+(path p1, path p2)
     return p1;
 }
 
+static inline path absolute(const path& p)
+{
+    return std::filesystem::absolute(p);
+}
+
 /**
  * Convert path object to byte string. On POSIX, paths natively are byte
  * strings so this is trivial. On Windows, paths natively are Unicode, so an
diff --git a/src/util/system.cpp b/src/util/system.cpp
index d6afe3b0f..cfacb2061 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -392,7 +392,7 @@ const fs::path& ArgsManager::GetBlocksDirPath() const
     if (!path.empty()) return path;
 
     if (IsArgSet("-blocksdir")) {
-        path = fs::absolute(GetArg("-blocksdir", ""));
+        path = fs::absolute(fs::PathFromString(GetArg("-blocksdir", "")));
         if (!fs::is_directory(path)) {
             path = "";
             return path;
@@ -419,7 +419,7 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
 
     std::string datadir = GetArg("-datadir", "");
     if (!datadir.empty()) {
-        path = fs::absolute(datadir);
+        path = fs::absolute(fs::PathFromString(datadir));
         if (!fs::is_directory(path)) {
             path = "";
             return path;
@@ -799,7 +799,7 @@ fs::path GetDefaultDataDir()
 bool CheckDataDirOption()
 {
     std::string datadir = gArgs.GetArg("-datadir", "");
-    return datadir.empty() || fs::is_directory(fs::absolute(datadir));
+    return datadir.empty() || fs::is_directory(fs::absolute(fs::PathFromString(datadir)));
 }
 
 fs::path GetConfigFile(const std::string& confPath)
```

But I'm not sure whether the diff above should belong to this PR.

hebasto commented:
I've managed to make functional tests pass on Windows (MSVC build) in the [combined branch](https://github.com/hebasto/bitcoin/commits/pr22937-0914-COMBINED).

And the following [patch](https://github.com/hebasto/bitcoin/commit/90c02435d1d92502afc354efe02ebbcad8f58548) was required:
```diff
--- a/src/addrdb.cpp
+++ b/src/addrdb.cpp
@@ -194,7 +194,7 @@ std::optional<bilingual_str> LoadAddrman(const std::vector<bool>& asmap, const A
     } catch (const DbNotFoundError&) {
         // Addrman can be in an inconsistent state after failure, reset it
         addrman = std::make_unique<CAddrMan>(asmap, /* deterministic */ false, /* consistency_check_ratio */ check_addrman);
-        LogPrintf("Creating peers.dat because the file was not found (%s)\n", path_addr);
+        LogPrintf("Creating peers.dat because the file was not found (%s)\n", fs::PathToString(path_addr));
         DumpPeerAddresses(args, *addrman);
     } catch (const std::exception& e) {
         addrman = nullptr;
```

Could such changes be forced by a compiler? _(some discussion [above](https://github.com/bitcoin/bitcoin/pull/22937#discussion_r706015262))_

laanwj commented:
It's so incredibly stupid that basic, straightforward, conversion functions like these are unsafe and we need to write our own replacements. Sorry, more of a C++ rant than anything else… along with the locale handling disaster it seems like an intentional minefield sometimes.

ryanofsky commented:
Thank you @hebasto for debugging and fixing this! 🙇 

Updated 650470227f9312b627b4bee3f85540f2e00f8bd7 -> 127cce3ac9d0aab6401eafbe8d4f0d2400f1ccf4 ([`pr/u8path.5`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.5) -> [`pr/u8path.6`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.5..pr/u8path.6)) with these fixes

re: https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-919610206

> I've managed to make functional tests pass on Windows (MSVC build) in the [combined branch](https://github.com/hebasto/bitcoin/commits/pr22937-0914-COMBINED).
> ...
> Could such changes be forced by a compiler?

Yes, added tinyformat override so there will be link errors.

ryanofsky commented:
Updated 127cce3ac9d0aab6401eafbe8d4f0d2400f1ccf4 -> 552e87e342a670808a227653a52134661400afcc ([`pr/u8path.6`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.6) -> [`pr/u8path.7`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.6..pr/u8path.7)) fixing unintended formatting changes caused by adding PathToString calls, which made some functional tests fail

kiminuo commented:
I did try to search for `.string()` and I got these occurrences:

<details>
  <summary>Search result</summary>

```
19 results - 11 files

src\dbwrapper.cpp:
  117  CDBWrapper::CDBWrapper(const fs::path& path, size_t nCacheSize, bool fMemory, bool fWipe, bool obfuscate)
  118:     : m_name{path.stem().string()}
  119  {

src\init.cpp:
  1097                    "also be data loss if bitcoin is started while in a temporary directory.\n",
  1098:                   args.GetArg("-datadir", ""), fs::current_path().string());
  1099      }

src\node\blockstorage.cpp:
  70          if (fs::is_regular_file(*it) &&
  71:             it->path().filename().string().length() == 12 &&
  72:             it->path().filename().string().substr(8,4) == ".dat")
  73          {
  74:             if (it->path().filename().string().substr(0, 3) == "blk") {
  75:                 mapBlockFiles[it->path().filename().string().substr(3, 5)] = it->path();
  76:             } else if (it->path().filename().string().substr(0, 3) == "rev") {
  77                  remove(it->path());

src\test\validation_chainstatemanager_tests.cpp:
  186      BOOST_TEST_MESSAGE(
  187:         "Wrote UTXO snapshot to " << snapshot_path.make_preferred().string() << ": " << result.write());
  188  

src\util\system.cpp:
  121      LOCK(cs_dir_locks);
  122:     dir_locks.erase((directory / lockfile_name).string());
  123  }

  246      auto result = path;
  247:     while (result.filename().string() == ".") {
  248          result = result.parent_path();

src\wallet\bdb.cpp:
  831          LOCK(cs_db); // Lock env.m_databases until insert in BerkeleyDatabase constructor
  832:         std::string data_filename = data_file.filename().string();
  833          std::shared_ptr<BerkeleyEnvironment> env = GetBerkeleyEnv(data_file.parent_path());
  834          if (env->m_databases.count(data_filename)) {
  835:             error = Untranslated(strprintf("Refusing to load database. Data file '%s' is already loaded.", (env->Directory() / data_filename).string()));
  836              status = DatabaseStatus::FAILED_ALREADY_LOADED;

src\wallet\bdb.h:
  143      /** Return path to main database filename */
  144:     std::string Filename() override { return (env->Directory() / strFile).string(); }
  145  

src\wallet\db.cpp:
  22                  it.no_push();
  23:                 LogPrintf("%s: %s %s -- skipping.\n", __func__, ec.message(), it->path().string());
  24              } else {
  25:                 LogPrintf("%s: %s %s\n", __func__, ec.message(), it->path().string());
  26              }

  53          } catch (const std::exception& e) {
  54:             LogPrintf("%s: Error scanning %s: %s\n", __func__, it->path().string(), e.what());
  55              it.no_push();

src\wallet\test\db_tests.cpp:
  18      fs::path data_file = BDBDataFile(path);
  19:     database_filename = data_file.filename().string();
  20      return GetBerkeleyEnv(data_file.parent_path());

src\wallet\test\wallet_tests.cpp:
  260  
  261:     std::string backup_file = (gArgs.GetDataDirNet() / "wallet.backup").string();
  262  
```
</details>

... and I wonder whether cases like the following ones (and probably others) should use the new safe API functions (i.e. `fs::PathToString`):

* https://github.com/ryanofsky/bitcoin/blob/552e87e342a670808a227653a52134661400afcc/src/init.cpp#L1098
* https://github.com/ryanofsky/bitcoin/blob/552e87e342a670808a227653a52134661400afcc/src/wallet/test/wallet_tests.cpp#L261

Please correct me if I'm wrong but my understanding is that `fs::PathToString` (and other helper functions) should be used as much as possible (and if not then it should be probably commented). Am I missing something?

Thanks for an explanation!

ryanofsky commented:
> I did try to search for `.string()` and I got these occurrences:

Thanks! Great idea.

> Please correct me if I'm wrong but my understanding is that `fs::PathToString` (and other helper functions) should be used as much as possible (and if not then it should be probably commented). Am I missing something?

No that's right. I just didn't think to search for these, and the tests that add special characters to directory names might not have caught some of these instances because they were only encoding filenames.

> Thanks for an explanation!

Thank you! Updated 552e87e342a670808a227653a52134661400afcc -> 8026743822c52b723c7ce1dd322fd32dfc23a79b ([`pr/u8path.7`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.7) -> [`pr/u8path.8`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.7..pr/u8path.8)) with replacements

ryanofsky commented:
Rebased a1b9c0f2a557dce39ae83f1738b7668c586f376e -> 8a2f979ea78c116dc586da9d18ef1697b297ef01 ([`pr/u8path.11`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.11) -> [`pr/u8path.12`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.11-rebase..pr/u8path.12)) due to conflict with #19806

hebasto commented on src/fs.h:
```diff
@@ -15,7 +15,122 @@
+// Allow explicit quoted stream I/O.
+static inline auto quoted(const std::string& s)
+{
+    return boost::io::quoted(s, '&');
```
8a2f979ea78c116dc586da9d18ef1697b297ef01

It is unfortunate that we have to do this.

Anyway, using of `boost::io::quoted` requires to include headers as follows:
```diff
diff --git a/src/fs.h b/src/fs.h
index e15dc3051..56934ea29 100644
--- a/src/fs.h
+++ b/src/fs.h
@@ -14,6 +14,13 @@
 #include <boost/filesystem.hpp>
 #include <boost/filesystem/fstream.hpp>
 
+#include <boost/version.hpp>
+#if BOOST_VERSION >= 107300
+#include <boost/io/quoted.hpp>
+#else
+#include <boost/io/detail/quoted_manip.hpp>
+#endif
+
 /** Filesystem operations and types */
 namespace fs {
 
diff --git a/test/lint/lint-includes.sh b/test/lint/lint-includes.sh
index bf7aeb5b4..974da45b9 100755
--- a/test/lint/lint-includes.sh
+++ b/test/lint/lint-includes.sh
@@ -56,6 +56,8 @@ EXPECTED_BOOST_INCLUDES=(
     boost/date_time/posix_time/posix_time.hpp
     boost/filesystem.hpp
     boost/filesystem/fstream.hpp
+    boost/io/detail/quoted_manip.hpp
+    boost/io/quoted.hpp
     boost/multi_index/hashed_index.hpp
     boost/multi_index/ordered_index.hpp
     boost/multi_index/sequenced_index.hpp
@@ -65,6 +67,7 @@ EXPECTED_BOOST_INCLUDES=(
     boost/signals2/optional_last_value.hpp
     boost/signals2/signal.hpp
     boost/test/unit_test.hpp
+    boost/version.hpp
 )
 
 for BOOST_INCLUDE in $(git grep '^#include <boost/' -- "*.cpp" "*.h" | cut -f2 -d: | cut -f2 -d'<' | cut -f1 -d'>' | sort -u); do
```

hebasto replied:
> > Anyway, using of boost::io::quoted requires to include headers as follows:
> 
> I don't think this is true...

It was required to compile the combined branch -- this PR + #20744.

ryanofsky replied:
> It was required to compile the combined branch -- this PR + #20744.

Maybe we can move the discussion to #20744 if there are continuing issues there but I think #20744 should not be calling boost::io::quoted but should switch to std::quoted instead. We can also choose to get rid of this fs::quoted function entirely in a new PR between this PR and #20744 if we are willing to changing logging behavior a little bit.

hebasto replied:
Agree.

hebasto reviewed (commented):
I tested the combined branch (this PR + #20744) on top of the master (16ccb3a1cd9125eb24a5b45a98099ff98660767a), and found another implicit type conversion case which could be fixed with the following patch:

```diff
diff --git a/src/fs.h b/src/fs.h
index e15dc3051..9d611f44f 100644
--- a/src/fs.h
+++ b/src/fs.h
@@ -71,6 +71,11 @@ static inline path system_complete(const path& p)
     return boost::filesystem::system_complete(p);
 }
 
+static inline bool exists(const path& p)
+{
+    return boost::filesystem::exists(p);
+}
+
 // Allow explicit quoted stream I/O.
 static inline auto quoted(const std::string& s)
 {
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 4e92889aa..c255a950c 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -2842,7 +2842,7 @@ static RPCHelpMan restorewallet()
 
     std::string backup_file = request.params[1].get_str();
 
-    if (!fs::exists(backup_file)) {
+    if (!fs::exists(fs::PathFromString(backup_file))) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Backup file does not exist");
     }
 
```

---

UPDATE:

[`fs::copy_file`](https://github.com/bitcoin/bitcoin/blob/09cb5ec6c8b4468ee3e8c2b3e1a9075907e5c84d/src/wallet/rpcwallet.cpp#L2878) arguments also must be protected from implicit type conversion.

The updated patch for `src/wallet/rpcwallet.cpp` looks as following:
```diff
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -2840,7 +2840,7 @@ static RPCHelpMan restorewallet()
 
     WalletContext& context = EnsureWalletContext(request.context);
 
-    std::string backup_file = request.params[1].get_str();
+    auto backup_file = fs::PathFromString(request.params[1].get_str());
 
     if (!fs::exists(backup_file)) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Backup file does not exist");
```

ryanofsky commented:
> I tested the combined branch (this PR + #20744) on top of the master ([16ccb3a](https://github.com/bitcoin/bitcoin/commit/16ccb3a1cd9125eb24a5b45a98099ff98660767a)), and found another implicit type conversion case which could be fixed with the following patch:

Another string bites the dust. Thanks!

Updated 8a2f979ea78c116dc586da9d18ef1697b297ef01 -> 371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68 ([`pr/u8path.12`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.12) -> [`pr/u8path.13`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.12..pr/u8path.13)) with suggested backup file fix and `fs::exists` string protection.

hebasto reviewed (approved):
ACK 371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68

kiminuo commented:
@ryanofsky I was looking around and I have found this function in `system.cpp`:

```cpp
bool RenameOver(fs::path src, fs::path dest)
{
#ifdef WIN32
    return MoveFileExW(src.wstring().c_str(), dest.wstring().c_str(),
                       MOVEFILE_REPLACE_EXISTING) != 0;
#else
    int rc = std::rename(src.c_str(), dest.c_str());
    return (rc == 0);
#endif /* WIN32 */
}
```
(https://github.com/ryanofsky/bitcoin/blob/371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68/src/util/system.cpp#L1064-L1073)

Is `.wstring().c_str()` correct? I would expect your `fs::PathToString` fuction to be used here too ... or maybe I would expect `std::filesystem::rename` here to avoid `#ifdef` completely. (A lot of guessing on my side.)

ryanofsky commented:
> Is `.wstring().c_str()` correct? I would expect your `fs::PathToString` fuction to be used here too

This is actually correct and calling PathToString would be wrong here. The idea behind this code is to avoid any conversion to narrow strings. The path object on windows natively uses a wide string, wstring() returns it, and 
MoveFileExW expects it. So there is no need for this code to change, even though it could be replaced by std::rename anytime I think if someone wanted to do that in a different PR

kiminuo commented:
https://github.com/ryanofsky/bitcoin/blob/371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68/src/i2p.cpp#L331 - Should `fs::PathToString` be used here?

ryanofsky commented:
> https://github.com/ryanofsky/bitcoin/blob/371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68/src/i2p.cpp#L331 - Should `fs::PathToString` be used here?

Wow, great find! This should have triggered a build error, and it was a nightmare to figure out what was allowing this to succeed, but it turned out I was overloading the tinyformat `formatValue` function in an imperfect way that caused the overload to be ignored if `fs.h` was included after `tinyformat.h`, and only used when `fs.h` was included *before* `tinyformat.h`. This worked most places, but was broken in the i2p source file. I fixed this by switching `formatValue` declarations to be template specializations instead of overloads to prevent this fragility and make the check actually reliable.

---

Updated 371b06d77a53cd1e9ac10c9f2b0ff9c56a2d8d68 -> b95db7a0393a4da8954f0a63371a673231c36b6e ([`pr/u8path.13`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.13) -> [`pr/u8path.14`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.13..pr/u8path.14)) to prevent a path with a locale-specific encoding from being formatted in an I2P error message on windows after #20744, and to fix a `valueFormat` overload bug that allowed this bad formatting to go undetected.

hebasto commented on src/fs.h:
```diff
@@ -103,4 +226,11 @@ namespace fsbridge {
 #endif // WIN32 && __GLIBCXX__
 };
 
+// Disallow implicit std::string conversion for tinyformat to avoid
+// locale-dependent encoding on windows.
+namespace tinyformat {
+template<> inline void formatValue(std::ostream&, const char*, const char*, int, const boost::filesystem::path&);
+template<> inline void formatValue(std::ostream&, const char*, const char*, int, const fs::path&);
+} // namespace tinyformat
```
A note (not an issue).
If code contains a `formatValue` call which assumes implicit `std::string` conversion, an error will be generated by a _linker_, not a compiler.

hebasto replied:
```suggestion
// Disallow implicit std::string conversion for tinyformat to avoid
// locale-dependent encoding on windows.
namespace tinyformat {
template<> inline void formatValue(std::ostream&, const char*, const char*, int, const fs::path&) = delete;
} // namespace tinyformat
```

to force a compiler to report errors?

Also the explicit template specialization with the `boost::filesystem::path` type seems redundant, no?

ryanofsky replied:
> to force a compiler to report errors?

Wow, I had no idea you could delete nonmember functions. It is much nicer to have this caught by the compiler. Will update.

> Also the explicit template specialization with the `boost::filesystem::path` type seems redundant, no?

This is just to things cases like `strformat("%s", path1 / path2)` or whatever else might return a `std::filesystem::path` instead of an `fs::path`

hebasto replied:
> This is just to catch cases like `strformat("%s", path1 / path2)` or whatever else might return a `std::filesystem::path` instead of an `fs::path`

We do not use the same assumption for other functions. Why just here?

ryanofsky replied:
> We do not use the same assumption for other functions. Why just here?

If there are other places where it's easy to catch more encoding problems, I would be happy to add overloads. This case does work a little differently from other cases because it is template specialization not a function overload. A function overload will match when a subclass instance is passed to a parent class parameter, while template specialization will not match because type deduction won't deduce a parent type from a child type.

hebasto replied:
Thanks for an explanation!

ryanofsky replied:
If we did want to be more vigilant about possible encoding bugs, we could have a followup PR that drops public inheritance `class path : public boost::filesystem::path` and wraps a private member instead, or uses private inheritance, to make the class a whitelist of allowed functions instead of a blacklist of disallowed functions. The changes required would be a superset of changes already in this PR, and it's unclear whether any bugs this might catch would be severe enough to justify the extra maintenance burden, so I didn't try this here. But it is something that could be done later.

hebasto reviewed (approved):
re-ACK b95db7a0393a4da8954f0a63371a673231c36b6e

I grepped `git grep -n -e 'strprintf.*fs::'`, and wonder why do we use `fs::quoted` to wrap `fs::PathToString` only in certain cases?

ryanofsky reviewed (commented):
> I grepped `git grep -n -e 'strprintf.*fs::'`, and wonder why do we use `fs::quoted` to wrap `fs::PathToString` only in certain cases?

I'm just preserving previous behavior in all cases. We previously sometimes printed paths with quotes, sometimes printed paths without quotes, and this PR is keeping it all the same so nothing breaks (tests, monitoring tools, etc)

ryanofsky commented:
Updated b95db7a0393a4da8954f0a63371a673231c36b6e -> bf9197c563cc62847a331df38c0a612433ea8beb ([`pr/u8path.14`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.14) -> [`pr/u8path.15`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.14..pr/u8path.15)) taking hebasto's suggestion and making the tinyformat link errors into compile errors

hebasto reviewed (approved):
re-ACK bf9197c563cc62847a331df38c0a612433ea8beb, only suggested changes since my [previous](https://github.com/bitcoin/bitcoin/pull/22937#pullrequestreview-765240532) review.

kiminuo commented on src/test/fs_tests.cpp:
```diff
@@ -11,6 +11,26 @@
+    BOOST_CHECK_EQUAL(fs::PathToString(fs::u8path(u8_str)), u8_str);
+#ifndef WIN32
+    // On non-windows systems, verify that non-UTF8 strings can be round
+    // tripped successfully. On windows, file paths are natively represented as
+    // unicode strings, so an invalid UTF-8 sequence that does not map to any
```
nit: I find the sentence "On windows, file paths are natively represented as unicode strings, [..]" slightly confusing because of that "unicode strings". My immediate understanding is something like "a sequence of bytes in either UTF-8, UTF-16, etc. encoding" which makes me guess for no good reason :)

According to https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160#syntax (and your [comment](https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-928112127) too), it holds that *Windows uses a null-terminated sequence of wchar_t, encoded as UTF-16 (one or more elements for each character).*

So I would expect something like "On Windows, file paths are natively represented by wchar_t sequences, encoded as UTF-16 strings, [..]" 

Hopefully, I get it right.

ryanofsky replied:
> nit: I find the sentence "On windows, file paths are natively represented as unicode strings, [..]" slightly confusing because of that "unicode strings". 

Thanks rewrote this and replaced "unicode strings" with "unicode". The character types and encodings that windows uses internally are not really relevant here. The point is just that paths on windows are unicode and have to be encoded as bytes, while paths on posix are already bytes and do not get encoded.

> My immediate understanding is something like "a sequence of bytes in either UTF-8, UTF-16, etc. encoding" which makes me guess for no good reason :)

This isn't actually true, see https://github.com/rust-lang/rust/issues/12056 as a starting point if you are really curious about this. But this comment is just explaining why the test is skipped on windows.

> According to https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160#syntax (and your [comment](https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-928112127) too), it holds that _Windows uses a null-terminated sequence of wchar_t, encoded as UTF-16 (one or more elements for each character)._

This seems to be a simplification, they are wchar_t, but not strictly valid utf-16. Not too relevant to the API in this PR, though, which is just trying to preserve current behavior and use a safe string encoding.

kiminuo commented on src/test/fs_tests.cpp:
```diff
@@ -11,6 +11,26 @@
+BOOST_AUTO_TEST_CASE(fsbridge_pathtostring)
+{
+    std::string u8_str = "fs_tests_₿_🏃";
+    BOOST_CHECK_EQUAL(fs::PathToString(fs::PathFromString(u8_str)), u8_str);
```
I wonder if it is worth adding a test for [.stem()](https://en.cppreference.com/w/cpp/filesystem/path/stem) method too. We use it in [dbwrapper.cpp](https://github.com/bitcoin/bitcoin/pull/22937/files#diff-44bef8fa93d168379ef331d9e11d333be06ce98a8f37ec65f8a987846b90a151R118). 

The test can be simply:

```cpp
BOOST_CHECK_EQUAL(fs::PathToString(fs::PathFromString(u8_str).stem()), u8_str);
```

or something like

```cpp
std::string test_filename = "fs_tests_₿_🏃.dat";
std::string expected_stem = "fs_tests_₿_🏃";
BOOST_CHECK_EQUAL(fs::PathToString(fs::PathFromString(u8_filename).stem()), expected_stem);
```

The reasoning behind this is to have a test that tests: string -> path -> modified path -> string. So we would be more sure that there is no std::filesystem implementation that behaves in a surprising way (I'm thinking about Android std::filesystem implementation that seems to be quite new.)

Anyway, it's just an idea from a category "maybe nice to have" and not "must have". Feel free to ignore.

ryanofsky reviewed (commented):
Rebased bf9197c563cc62847a331df38c0a612433ea8beb -> 2f4a115f4aa394323118e0b9d3569a089e280360 ([`pr/u8path.15`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.15) -> [`pr/u8path.16`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.15-rebase..pr/u8path.16)) after #23136 to include fix for `wallet_basic.py` `Fee of 0.00000255 BTC too low!` error https://cirrus-ci.com/task/4686259315539968?logs=ci#L30, and also including new review suggestions

ryanofsky commented:
Rebased 2f4a115f4aa394323118e0b9d3569a089e280360 -> 6544ea5035268025207d2402db2f7d90fde947a6 ([`pr/u8path.16`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.16) -> [`pr/u8path.17`](https://github.com/ryanofsky/bitcoin/commits/pr/u8path.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/u8path.16-rebase..pr/u8path.17)) due to conflict with #22950

hebasto reviewed (approved):
re-ACK 6544ea5035268025207d2402db2f7d90fde947a6, only added `fsbridge_stem` test case, updated comment, and rebased since my [previous](https://github.com/bitcoin/bitcoin/pull/22937#pullrequestreview-765503126) review. Verified with the following command:
```
$ git range-diff master bf9197c563cc62847a331df38c0a612433ea8beb 6544ea5035268025207d2402db2f7d90fde947a6
```

kiminuo commented:
ACK 6544ea5035268025207d2402db2f7d90fde947a6

laanwj commented:
Code review ACK 6544ea5035268025207d2402db2f7d90fde947a6
